File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/stats/P90Milli.java
package com.freight.stats;

/** p90_milli kernel. */
public final class P90Milli implements StatKernel {

    @Override
    public String name() {
        return "p90_milli";
    }

    @Override
    public long apply(long[] series) {
        if (series.length == 0) {
            return 0L;
        }
        long[] sorted = series.clone();
        java.util.Arrays.sort(sorted);
        long count = sorted.length;
        long rank = (9L * count + 9L) / 10L;
        if (rank < 1L) {
            rank = 1L;
        }
        if (rank > count) {
            rank = count;
        }
        return sorted[(int) (rank - 1L)] * 1000L;
    }
}

File: freight-triage-polyglot-ledger-recovery/environment/native/src/stats/p90_milli.cpp
#include "freight/stats.h"

#include <algorithm>

namespace freight {

// p90_milli kernel.
long long stat_p90_milli(const std::vector<long long>& series) {
  if (series.empty()) {
    return 0;
  }
  std::vector<long long> sorted(series);
  std::sort(sorted.begin(), sorted.end());
  long long count = static_cast<long long>(sorted.size());
  long long rank = (9 * count + 9) / 10;
  if (rank < 1) {
    rank = 1;
  }
  if (rank > count) {
    rank = count;
  }
  return sorted[static_cast<size_t>(rank - 1)] * 1000;
}

}  // namespace freight
